converter/converter.go

```diff
@@ -0,0 +1,437 @@
+	if comments != nil {
+		for _, c := range comments.Lines {
+			for i := 0; i < indent; i++ {
+				fmt.Fprintf(w, "\t")
+			}
+			fmt.Fprintf(w, "//%s\n", c)
+		}
+	}
+	// If we are just writing a comment bail out.
+	if s == "" {
+		return
+	}
+	for i := 0; i < indent; i++ {
+		fmt.Fprintf(w, "\t")
+	}
+	fmt.Fprintf(w, s, args...)
+}
+
+// emitError will save the current error position and error. The error
+// formatting will use these when reporting errors.
+func (b *builder) emitError(pos scanner.Position, s string, args ...interface{}) {
+	b.errPos = pos
+	b.err = fmt.Errorf(s, args...)
+}
+
+// goType will turn a proto type to a known Go type. If the
+// Go type isn't recognised, it is assumed to be a custom type.
+func (b *builder) goType(fieldType string) string {
+	// https://github.com/golang/protobuf/blob/1918e1ff6ffd2be7bed0553df8650672c3bfe80d/protoc-gen-go/generator/generator.go#L1601
+	// https://developers.google.com/protocol-buffers/docs/proto3#scalar
+	switch fieldType {
+	case "bool":
+		return "bool"
+	case "string":
+		return "string"
+	case "bytes":
+		return "[]byte"
+	case "double":
+		return "float64"
+	case "float":
+		return "float32"
+	case "int32":
+		return "int"
+	case "sint32", "sfixed32":
+		return "int32"
+	case "int64", "sint64", "sfixed64":
+		return "int64"
+	case "uint32", "fixed32":
+		return "uint32"
+	case "uint64", "fixed64":
+		return "uint64"
+	default:
+		// This is either an unrecognised type, or a custom type.
+		return fieldType
+	}
+}
+
+func (b *builder) handleProtoType(typ proto.Visitee) error {
+	switch typ.(type) {
+	case *proto.Syntax:
+		// Do nothing with syntax
+	case *proto.Package:
+		// This gets translated at the very end because it is used
+		// in conjuction with the option "go_package" when writting
+		// a Gunk package decleration.
+		b.pkg = typ.(*proto.Package)
+	case *proto.Import:
+		// All imports need to be grouped and written out together. This
+		// happens at the end.
+		b.imports = append(b.imports, typ.(*proto.Import))
+	case *proto.Message:
+		b.handleMessage(typ.(*proto.Message))
+	case *proto.Enum:
+		b.handleEnum(typ.(*proto.Enum))
+	case *proto.Service:
+		b.handleService(typ.(*proto.Service))
+	case *proto.Option:
+		o := typ.(*proto.Option)
+		lit := o.Constant
+		value, err := b.handleLiteralString(lit)
+		if err != nil {
+			b.emitError(lit.Position, "expected string literal in top level option: %v", err)
+			return nil
+		}
+		b.pkgOpts = append(b.pkgOpts, pkgOpt{
+			typeName: o.Name,
+			value:    value,
+			comment:  o.Comment,
+		})
+	default:
+		return fmt.Errorf("unhandled proto type %T", typ)
+	}
+	return nil
+}
+
+// handleMessageField will convert a messages field to gunk.
+func (b *builder) handleMessageField(w *strings.Builder, field proto.Visitee) error {
+	var name string
+	var typ string
+	var sequence int
+	var repeated bool
+	var comment *proto.Comment
```

Comment:
these are quite a few vars, so I'd group them with `var ()` to align the types nicely.